### DISCUSSION TITLE
Programmatically add product variants via API

### DISCUSSION THREAD

**Discussion by de-raaf-media** [duncanmcclean/simple-commerce#932] [Help] [answered]
Hi there!

I am working on an auto product importer that runs every night for CSV files containing products. I am running into a problem:

I tried to Programmatically add product variants via API like so. 

$product->set('product_variants', $product_variants_array);

But the variants are not added. However, if I use a double underscore for the key name (or any other random string) it is added. And if I then in de database rename product__variants to product_variants (single underscore) it DOES work, and the variants appear in the CP.

$product->set('product__variants', $product_variants_array);

Is there a reason for this behavior? I noticed something similar when I tried where I tried to execute $product->set('site') where $product->set('somerandomstring') works fine.

Can I not overwrite default keys maybe? For now after finishing my product import, I use a replace query to change **product__variants** to **product_variants** (and dan it all works) but this is not ideal and I am trying to understand why this is happening ;-)

I am a bit stuck here. Would greatly appreciate some help.

Mark

**Comment by duncanmcclean** [comment, marked as answer]
Product Variants are special - you can't set them in the same way you can any other field. You need to use the specific `productVariants` method.

When you set the variants, you need to provide an array for both the options & the individual variants. Make sure you have a variant for each of the combinations.

You can set the options & variants like this:

```php
$product->productVariants([
  'variants' => [
    ['name' => 'Colour', 'values' => ['Red', 'Green', 'Blue']],
    ['name' => 'Size', 'values' => ['Small', 'Medium']]
  ],
  'options' => [
    [
      'key' => 'Red_Small',
      'variant' => 'Red, Small',
      'price' => 1599,
    ],
    [
      'key' => 'Red_Medium',
      'variant' => 'Red, Medium',
      'price' => 2950,
    ],
    // Green_Small
    // Green_Medium
    // Blue_Small
    // Blue_Medium
  ],
]);
```

Hopefully that's helpful, at least somewhat.

**Reply by de-raaf-media** [reply]
That is very helpful indeed! 

My Array is valid because changing __ to _ in my example does result in valid variants in my product.

I think the only thing I have to do is change `$product->set('product_variants' ...` to` $product->productVariants(...`

Thanks!

**Reply by de-raaf-media** [reply]
Slowly getting there with my product importer. However. I just discovered some behavior I cannot explain. If I add te product via the API the tax category in data contains a JSON object. I would expect a string there. Ik searched through the Simple Commerce code for a solution but no luck so far.

Instead of:
"tax_category": {"id": "default", "name": "Default", "description": "Will be used for all products where a category has not been assigned."},

I would expect (because if I add a tax_category field to my blueprint and manually add a new product this is the result):
"tax_category": {"default"},

So far I tried:

- $product->set('tax_category', 'default');
- $product->taxCategory('default');
- not specifying the tax_category at all

Edit: The problem is that I cannot edit the product if the tax category object is added. This seems to have something to do with the tax category id.

Mark

**Reply by duncanmcclean** [reply]
Tax Category is another special one... you need to use the [`->taxCategory()`](https://github.com/duncanmcclean/simple-commerce/blob/5.x/src/Products/Product.php#L81C21-L81C32) method to set it. 

`$product->taxCategory('default')` should work as long as you have a tax category with a handle of `default`. 

In terms of getting it from the API end, I _would_ expect it to return a JSON object with keys like that because the "Tax Category" fieldtype augments to the 'full' tax category, in the same way if you had an entries fieldtype, it would augment to the 'full' entry.

**Reply by de-raaf-media** [reply]
Hi Duncan, I did use the taxCategory method. The only Issue I have with this is that adding a product via the API wil generate a JSON object. Adding a product via the CP will generate a string. 

The result is that I am unable to edit a product in the CP that was added via the API.

**Reply by duncanmcclean** [reply]
Sorry, I've been busy over the last couple of weeks. 

This sounds like a bug though... are you able to open a GitHub Issue with the code you're attempting to use and what about it isn't working. I'll try and take a look this week if I can. Thanks!

**Reply by de-raaf-media** [reply]
Okay, thanks. I'll come back to this.
